### DISCUSSION TITLE
Feature Request: IDOM URL Router

### DISCUSSION THREAD

**Discussion by Archmonger** [idom-team/idom#508] [Ideas]
Currently it is very awkward to build SPAs within IDOM. There needs to be some kind of URL router construct.

**Comment by tclasen** [comment]
Some client based router like react-router or what nextjs has would be ideal. That would allow for a page refresh or sharing a link to get back to the same viewable state.

**Reply by rmorshea** [reply]
The development of this feature is under way. The foundational work has already been completed (i.e. `idom.backend.*.use_location` hooks). I'd expect that we might have an initial implementation ready to try out before July.

**Comment by acivitillo** [comment]
Yes, this one is critical to building SPAs in IDOM, we are using `use_state` to hide/show components but it's awkward to use

**Reply by rmorshea** [reply]
The development of this feature is under way. The foundational work has already been completed (i.e. `idom.backend.*.use_location` hooks). I'd expect that we might have an initial implementation ready to try out before July.

**Comment by rmorshea** [comment]
A very early implementation can be found here: https://github.com/idom-team/idom-router

If you decide to use it now, keep in mind that there are probably bugs and there will be upcoming changes the interface.
